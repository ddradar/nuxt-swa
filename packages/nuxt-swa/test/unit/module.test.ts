// @vitest-environment node
import { describe, expect, it } from 'vitest'

import {
  generateTypeDefinition,
  resolveAuthProviders,
  type ModuleOptions,
} from '~~/src/module'

describe('module.ts', () => {
  describe('resolveAuthProviders', () => {
    it.each([undefined, {}, { auth: {} }])(
      '(%o) returns ["aad", "github"]',
      config =>
        expect(resolveAuthProviders(config)).toStrictEqual(['aad', 'github'])
    )

    it.each([
      [{}, ['aad', 'github']],
      [
        { google: {}, twitter: { enabled: false } },
        ['aad', 'github', 'google'],
      ],
      [
        {
          azureActiveDirectory: { enabled: false },
          customOpenIdConnectProviders: { line: {}, yahoo: { enabled: false } },
        },
        ['github', 'line'],
      ],
    ])(
      '( azure.config.auth.identityProviders: %o ) returns %o',
      (identityProviders, expected) =>
        expect(
          resolveAuthProviders({ auth: { identityProviders } })
        ).toStrictEqual(expected)
    )
  })

  describe('generateTypeDefinition', () => {
    it.each([
      {
        authProviders: ['aad', 'github'],
        customRoles: [],
        dataApi: { rest: '/rest', graphql: '/graphql' },
      },
      {
        authProviders: [],
        customRoles: [],
        dataApi: false,
      },
      {
        authProviders: ['github', 'twitter', 'line'],
        customRoles: ['administrator'],
        dataApi: { rest: '/rest', graphql: '/graphql' },
      },
    ] satisfies ModuleOptions[])(
      '(%o) renders expected Type Definition',
      options => expect(generateTypeDefinition(options)).toMatchSnapshot()
    )
  })
})
