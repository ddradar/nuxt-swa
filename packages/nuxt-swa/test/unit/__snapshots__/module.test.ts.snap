// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`module.ts > generateTypeDefinition > ({ authProviders: [ 'aad', 'github' ], customRoles: [], dataApi: { rest: '/rest', graphql: '/graphql' } }) renders expected Type Definition 1`] = `
"// Generated by nuxt-swa
/**
 * User roles to use in your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/authentication-custom#manage-roles
 */
type UserRole = 'anonymous' | 'authenticated'

interface Claim {
  typ: string
  val: string
}

/**
 * User-identifiable information to your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/user-information?tabs=javascript#client-principal-data
 */
export interface ClientPrincipal {
  /** The name of the identity provider. */
  identityProvider: IdentityProvider
  /**
   * An Azure Static Web Apps-specific unique identifier for the user.
   * - The value is unique on a per-app basis. For instance, the same user returns a different \`userId\` value on a different Static Web Apps resource.
   * - The value persists for the lifetime of a user. If you delete and add the same user back to the app, a new \`userId\` is generated.
   */
  userId: string
  /**
   * Username or email address of the user.
   * Some providers return the user's email address, while others send the user handle.
   */
  userDetails: string
  /** An array of the user's assigned roles. */
  userRoles: UserRole[]
  /**
   * An array of claims returned by your custom authentication provider.
   * Only accessible in the direct-access endpoint (/.auth/me).
   */
  claims?: Claim[]
}

// Auth Feature
/**
 * Authentication Provider list to login your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/authentication-custom#configure-a-custom-identity-provider
 */
type IdentityProvider = 'aad' | 'github'

// Data API (preview) Feature
/**
 * Data API (REST) response
 * @see https://learn.microsoft.com/azure/data-api-builder/rest#result-set-format
 */
type RestResult<T> = { value: T[] }

/**
 * Data API (GraphQL) response
 * @see https://learn.microsoft.com/azure/data-api-builder/graphql#resultset-format
 */
type GraphQLResult<T> = { data: T }
"
`;

exports[`module.ts > generateTypeDefinition > ({ authProviders: [ 'github', 'twitter', 'line' ], customRoles: [ 'administrator' ], dataApi: { rest: '/rest', graphql: '/graphql' } }) renders expected Type Definition 1`] = `
"// Generated by nuxt-swa
/**
 * User roles to use in your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/authentication-custom#manage-roles
 */
type UserRole = 'anonymous' | 'authenticated' | 'administrator'

interface Claim {
  typ: string
  val: string
}

/**
 * User-identifiable information to your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/user-information?tabs=javascript#client-principal-data
 */
export interface ClientPrincipal {
  /** The name of the identity provider. */
  identityProvider: IdentityProvider
  /**
   * An Azure Static Web Apps-specific unique identifier for the user.
   * - The value is unique on a per-app basis. For instance, the same user returns a different \`userId\` value on a different Static Web Apps resource.
   * - The value persists for the lifetime of a user. If you delete and add the same user back to the app, a new \`userId\` is generated.
   */
  userId: string
  /**
   * Username or email address of the user.
   * Some providers return the user's email address, while others send the user handle.
   */
  userDetails: string
  /** An array of the user's assigned roles. */
  userRoles: UserRole[]
  /**
   * An array of claims returned by your custom authentication provider.
   * Only accessible in the direct-access endpoint (/.auth/me).
   */
  claims?: Claim[]
}

// Auth Feature
/**
 * Authentication Provider list to login your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/authentication-custom#configure-a-custom-identity-provider
 */
type IdentityProvider = 'github' | 'twitter' | 'line'

// Data API (preview) Feature
/**
 * Data API (REST) response
 * @see https://learn.microsoft.com/azure/data-api-builder/rest#result-set-format
 */
type RestResult<T> = { value: T[] }

/**
 * Data API (GraphQL) response
 * @see https://learn.microsoft.com/azure/data-api-builder/graphql#resultset-format
 */
type GraphQLResult<T> = { data: T }
"
`;

exports[`module.ts > generateTypeDefinition > ({ authProviders: [], customRoles: [], dataApi: false }) renders expected Type Definition 1`] = `
"// Generated by nuxt-swa
/**
 * User roles to use in your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/authentication-custom#manage-roles
 */
type UserRole = 'anonymous' | 'authenticated'

interface Claim {
  typ: string
  val: string
}

/**
 * User-identifiable information to your app.
 * @see https://learn.microsoft.com/azure/static-web-apps/user-information?tabs=javascript#client-principal-data
 */
export interface ClientPrincipal {
  /** The name of the identity provider. */
  identityProvider: IdentityProvider
  /**
   * An Azure Static Web Apps-specific unique identifier for the user.
   * - The value is unique on a per-app basis. For instance, the same user returns a different \`userId\` value on a different Static Web Apps resource.
   * - The value persists for the lifetime of a user. If you delete and add the same user back to the app, a new \`userId\` is generated.
   */
  userId: string
  /**
   * Username or email address of the user.
   * Some providers return the user's email address, while others send the user handle.
   */
  userDetails: string
  /** An array of the user's assigned roles. */
  userRoles: UserRole[]
  /**
   * An array of claims returned by your custom authentication provider.
   * Only accessible in the direct-access endpoint (/.auth/me).
   */
  claims?: Claim[]
}

// Auth Feature

// Data API (preview) Feature

"
`;
